src/WebJobs.Script.WebHost/Management/LinuxInstanceManager.cs

```diff
@@ -62,24 +98,28 @@ public bool StartAssignment(HostAssignmentContext context)
                 return false;
             }
 
-            if (context.IsWarmupRequest)
-            {
-                // Based on profiling download code jit-ing holds up cold start.
-                // Pre-jit to avoid paying the cost later.
-                Task.Run(async () => await DownloadWarmupAsync(context.GetRunFromPkgContext()));
-                return true;
-            }
-            else if (_assignmentContext == null)
+            return true;
+        }
+
+        private async Task AssignAsync(HostAssignmentContext assignmentContext)
+        {
+            if (assignmentContext.IsWarmupRequest)
             {
-                lock (_assignmentLock)
+                try
+                {
+                    await DownloadWarmupAsync(assignmentContext.GetRunFromPkgContext());
+                }
+                catch (Exception ex)
                 {
-                    if (_assignmentContext != null)
-                    {
-                        return _assignmentContext.Equals(context);
-                    }
-                    _assignmentContext = context;
+                    _logger.LogError(ex, "Warmup download failed");
+                    await _meshServiceClient.NotifyHealthEvent(ContainerHealthEventType.Warning, GetType(), "Warmup download failed");
+                    throw;
                 }
+                return;
+            }
 
+            try
+            {
                 _logger.LogInformation($"Starting Assignment. Cloud Name: {_environment.GetCloudName()}");
```

Comment:
Debating whether we should change this to take in a log argument. It would technically a break of this log statement.

Reply:
Also this is a standardized piece of telemetry, so we wouldn't want a different name for it than the standards